CapaModule to ProblemBlock [SE-397]

symbolist opened the pull request:
This PR is based on https://github.com/edx/edx-platform/pull/19071. To reduce risk and make review and testing easier, it brings in only the changes needed to merge `CapaModule` and `CapaDescriptor` into a single `ProblemBlock` class. Other required changes can be pulled in separately in subsequent PRs.

**Dependencies**: None

**Sandbox URL**:
LMS: https://pr20023.sandbox.opencraft.hosting
Studio: https://studio.pr19941.sandbox.opencraft.hosting

**Merge deadline**: None

**Notes**:

- Changes in behavior are not expected (though we still need to do testing to verify this).
- `CapaModule` and `CapaDescriptor` inherit from a number of classes which inherited from `XModule` or `XModuleDescriptor` though did not depend on them. For all these classes the methods were moved to a mixin which does not inherit from either and then these mixins have been added to `ProblemBlock` in the order which maintains MRO.

CapaDescriptor MRO | CapaModule MRO | ProblemBlock MRO | Notes
------------ | ------------- | ---------------- | ------------
<class 'xmodule.capa_module.CapaDescriptor'> | <class 'xmodule.capa_module.CapaModule'> | <class 'xmodule.capa_module.ProblemBlock'>
N/A | <class 'xmodule.capa_base.CapaMixin'> | <class 'xmodule.capa_base.CapaMixin'>
N/A | <class 'xblock.scorable.ScorableXBlockMixin'> | <class 'xblock.scorable.ScorableXBlockMixin'>
<class 'xmodule.capa_base.CapaFields'> | <class 'xmodule.capa_base.CapaFields'> | Not needed. | Merged into ProblemBlock.
<class 'xmodule.raw_module.RawDescriptor'> | N/A | <class 'xmodule.raw_module.RawMixin'> | All methods in RawDescriptor were moved to RawMixin.
<class 'xmodule.xml_module.XmlDescriptor'> | N/A | <class 'xmodule.xml_module.XmlMixin'> | All methods in XmlDescriptor were moved to XmlMixin.
<class 'xmodule.xml_module.XmlParserMixin'> | N/A | <class 'xmodule.xml_module.XmlParserMixin'>
<class 'xmodule.editing_module.XMLEditingDescriptor'> | N/A | Not needed. | All 3 attributes are being overridden in ProblemBlock.
<class 'xmodule.editing_module.EditingDescriptor'> | N/A | <class 'xmodule.editing_module.EditingMixin'> | All methods were moved from EditingDescriptor to the EditingMixin.
<class 'xmodule.editing_module.EditingFields'> | N/A | <class 'xmodule.editing_module.EditingFields'>
<class 'xmodule.mako_module.MakoModuleDescriptor'> | N/A | Not needed. | Only has one method `get_html()`
<class 'xmodule.mako_module.MakoTemplateBlockBase'> | N/A | <class 'xmodule.mako_module.MakoTemplateBlockBase'>
<class 'xmodule.x_module.XModuleDescriptor'> | N/A | <class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | Required methods were moved from the XModuleDescriptor to the mixin.
N/A  | <class 'xmodule.x_module.XModule'> | <class 'xmodule.x_module.XModuleToXBlockMixin'> | Required methods were moved from XModule to the mixin.
<class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'>
<class 'xmodule.x_module.ResourceTemplates'> | N/A | <class 'xmodule.x_module.ResourceTemplates'>
<class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'>
<class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> 
<class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'>
xblock.mixins | xblock.mixins | xblock.mixins

edx-webhook commented:
Thanks for the pull request, @symbolist! I've created [OSPR-3224](https://openedx.atlassian.net/browse/OSPR-3224) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

symbolist commented:
I closed the earlier PR https://github.com/edx/edx-platform/pull/19915 because it looked like the status-bot wasn't posting all updates after I accidentally closed it.

natabene commented:
@symbolist Thank you for your contribution. Please let me know once it is ready for our review, either product or engineering.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -40,7 +36,30 @@
 
-FEATURES = getattr(settings, 'FEATURES', {})
+
+class SHOWANSWER(object):
```
Moved over from `common/lib/xmodule/xmodule/capa_base_constants.py` to reduce file clutter.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -85,157 +104,11 @@ def default(self, obj):  # pylint: disable=method-hidden
         return json.JSONEncoder.default(self, obj)
 
 
-class CapaFields(object):
```
`CapaFields` has been merged into `ProblemBlock` without any changes.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -328,7 +201,7 @@ def new_lcp(self, state, text=None):
 
         capa_system = LoncapaSystem(
-            ajax_url=self.runtime.ajax_url,
+            ajax_url=self.ajax_url,
```
The one in `ModuleSystem` works only for `XModules` so added one in `XModuleToXBlockMixin`.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
+try:
+    FEATURES = getattr(settings, 'FEATURES', {})
+except ImproperlyConfigured:
+    FEATURES = {}
```
Ran into some issues with running xmodule tests.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
     """
+    Define the possible fields for a Capa problem
+    """
+    display_name = String(
```
The fields were moved in from the `CapaFields` class. The only changes are a couple of tweaks for quality issues.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
+@XBlock.needs('i18n')
+class ProblemBlock(
+        CapaMixin, RawMixin, XmlParserMixin, EditingMixin,
+        XModuleDescriptorToXBlockMixin, XModuleToXBlockMixin, HTMLSnippet, ResourceTemplates, XModuleMixin):
```
The order maintains the MRO of `CapaModule` and `CapaDescriptor`.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
+    show_in_read_only_mode = True
+    template_dir_name = 'problem'
+    mako_template = "widgets/problem-edit.html"
+    has_author_view = True
```
The attributes were moved up from their location on the top of the `CapaDescriptor` below without any changes.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
+
+    # The capa format specifies that what we call max_attempts in the code
+    # is the attribute `attempts`. This will do that conversion
+    metadata_translations = dict(XmlParserMixin.metadata_translations)
```
Since we are no longer inheriting from `RawDescriptor` changed this to `XmlParserMixin` where this attribute is being inherited in `RawDescriptor`.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
         ]
     }
 
-    js_module_name = "Problem"
```
This is used by `shim_xmodule_js()`. Have changed `shim_xmodule_js()` to accept the module name as a parameter and passed in from `student_view()`.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -139,36 +356,6 @@ def display_name_with_default(self):
-    template_dir_name = 'problem'
-    mako_template = "widgets/problem-edit.html"
-    js = {'js': [resource_string(__name__, 'js/src/problem/edit.js')]}
-    js_module_name = "MarkdownEditingDescriptor"
```
This is used by `shim_xmodule_js()`. Have changed `shim_xmodule_js()` to accept the module name as a parameter and passed in from `studio_view()`.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
+        )
+        add_webpack_to_fragment(fragment, 'ProblemBlockStudio')
+        shim_xmodule_js(fragment, 'MarkdownEditingDescriptor')
+        return fragment
```
This was initially implemented in `MakoTemplateBlockBase` but needed to call `add_webpack_to_fragment()` and `shim_xmodule_js()` with custom parameters so copied over.

symbolist replied:
Both `add_webpack_to_fragment()` and `shim_xmodule_js` are called in `openedx/core/lib/xblock_utils/__init__.py` but that works because the Module and Descriptor are two different classes. In the interests to making this more XBlock like, have moved calling these to in `student_view()` and `studio_view()`.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -180,7 +367,7 @@ def filter_templates(cls, template, course):
 
     def get_context(self):
-        _context = RawDescriptor.get_context(self)
+        _context = EditingMixin.get_context(self)
```
This came from `RawDescriptor > XMLEditingDescriptor > EditingDescriptor`. `EditingDescriptor` methods have been moved to `EditingMixin` (that `ProblemBlock` inherits from now instead) so we can call it directly.

symbolist commented on common/lib/xmodule/xmodule/static_content.py:
```diff
@@ -18,28 +18,37 @@
+# Should only be used for XModules being converted to XBlocks.
+XBLOCK_CLASSES = [
+    ProblemBlock,
+]
```
Decided to add support for XBlocks to `static_content.py`. If we need to change the setup, we can do it once all XModules have been converted over in one go instead of doing it for each separately.

symbolist commented on common/lib/xmodule/xmodule/static_content.py:
```diff
@@ -81,7 +90,7 @@ def _write_styles(selector, output_root, classes):
+        class_css = getattr(class_, css_attribute)()
         for filetype in ('sass', 'scss', 'css'):
             for idx, fragment in enumerate(class_css.get(filetype, [])):
                 css_fragments[idx, filetype, fragment].add(class_.__name__)
```
For XBlocks, the name would change but since the styles are scoped with view names like the following (in `/common/static/xmodule/descriptor|modules/css/_module-styles.scss`) everything should continue to work correctly:
```
.xmodule_edit.xmodule_ProblemBlock {
}
```

symbolist commented on common/lib/xmodule/xmodule/static_content.py:
```diff
@@ -127,12 +136,12 @@ def _write_js(output_root, classes):
         for filetype in ('coffee', 'js'):
             for idx, fragment in enumerate(module_js.get(filetype, [])):
-                fragment_owners[(idx + 1, filetype, fragment)].append(class_.__name__)
+                fragment_owners[(idx + 1, filetype, fragment)].append(getattr(class_, js_attribute + '_bundle_name')())
```
The name is used by `add_webpack_to_fragment()`.

symbolist commented on common/lib/xmodule/xmodule/tests/test_capa_module.py:
```diff
@@ -112,22 +112,22 @@ def create(cls, attempts=None, problem_state=None, correct=False, xml=None, over
 
-        system = get_test_system()
+        system = get_test_system(course_id=location.course_key)
+        system.user_is_staff = kwargs.get('user_is_staff', False)
```
This isn't needed but I had already made the change earlier.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -4,50 +4,267 @@
+        Return the student view.
+        """
+        fragment = Fragment(self.get_html())
+        add_webpack_to_fragment(fragment, 'ProblemBlockPreview')
```
TODO: Should change this to use `self.get_preview_view_js_bundle_name()`.

symbolist replied:
Actually the class at runtime has the mixins mixed in so need to hardcode this.

symbolist commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -324,6 +365,15 @@ def __init__(self, *args, **kwargs):
         super(XModuleMixin, self).__init__(*args, **kwargs)
+        self.init_only_for_xmodules_converted_to_xblocks()
+
+    def init_only_for_xmodules_converted_to_xblocks(self):
```
I know it is a hack but it works. Opinions welcome.

symbolist commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -831,8 +883,52 @@ def __delete__(self, instance):
+    Common code needed by XModule and XBlocks converted from XModules.
+    """
+    @property
+    def ajax_url(self):
```
This was added because the one in `ModuleSystem` only works for `XModules`.

symbolist commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -831,8 +883,52 @@ def __delete__(self, instance):
+        return self.runtime.handler_url(self, 'xmodule_handler', '', '').rstrip('/?')
+
+    @XBlock.handler
+    def xmodule_handler(self, request, suffix=None):
```
This was moved over from `XModule` without any changes.

symbolist commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -1160,6 +1184,51 @@ def export_to_xml(self, resource_fs):
+    module_class = XModule
+
+    # ============================= STRUCTURAL MANIPULATION ===================
+    def __init__(self, *args, **kwargs):
```
Moved without any changes.

symbolist commented on common/lib/xmodule/xmodule/xml_module.py:
```diff
@@ -512,7 +512,7 @@ def non_editable_metadata_fields(self):
 
 
-class XmlDescriptor(XmlParserMixin, XModuleDescriptor):  # pylint: disable=abstract-method
+class XmlMixin(XmlParserMixin):
```
We want all these methods for `ProblemBlock` without the `XModuleDescriptor` so moved them to a mixin.

symbolist commented on openedx/core/lib/xblock_utils/__init__.py:
```diff
@@ -108,7 +108,7 @@ def wrap_xblock(
     ]
 
-    if isinstance(block, (XModule, XModuleDescriptor)):
+    if isinstance(block, (XModule, XModuleDescriptor)) or getattr(block, 'uses_xmodule_styles_setup', False):
```
Using the same setup means we do not have to touch any styles or make any adjustments.

symbolist commented on openedx/core/lib/xblock_utils/__init__.py:
```diff
@@ -120,8 +120,10 @@ def wrap_xblock(
+    if isinstance(block, (XModule, XModuleDescriptor)):
         data['type'] = block.js_module_name
-        shim_xmodule_js(block, frag)
+        shim_xmodule_js(frag, block.js_module_name)
```
The `ProblemBlock` does load the javascript bundle and shim itself in `student_view()` and `studio_view()`. This wouldn't work because we have merged two classes into one.

symbolist commented on common/lib/xmodule/xmodule/tests/test_xblock_wrappers.py:
```diff
@@ -49,7 +49,8 @@
     AnnotatableDescriptor: [{}],
-    CapaDescriptor: [{}],
+    # TODO: How many of these checks are still needed?
+    # CapaDescriptor: [{}],
```
TODO: Still need to check this.

symbolist commented on common/lib/xmodule/xmodule/raw_module.py:
```diff
@@ -60,6 +59,14 @@ def definition_to_xml(self, resource_fs):
             raise SerializationError(self.location, msg)
 
 
+class RawDescriptor(RawMixin, XmlDescriptor, XMLEditingDescriptor):
```
We move the methods to `RawMixin` which we `ProblemBlock` can inherit. The required code from `XmlDescriptor` and `XMLEditingDescriptor` is inherited separately keeping the MRO same.

symbolist commented on common/lib/xmodule/xmodule/editing_module.py:
```diff
@@ -31,20 +31,24 @@ def non_editable_metadata_fields(self):
         return _context
 
 
+class EditingDescriptor(EditingMixin, MakoModuleDescriptor):
```
We move the attributes to `EditingMixin ` which `ProblemBlock` can inherit. The required code from `MakoModuleDescriptor ` is inherited separately keeping the MRO same.

symbolist commented on common/lib/xmodule/xmodule/editing_module.py:
```diff
@@ -31,20 +31,24 @@ def non_editable_metadata_fields(self):
     # set and then add our additional fields. Trying to keep it DRY.
     def get_context(self):
-        _context = MakoModuleDescriptor.get_context(self)
+        _context = MakoTemplateBlockBase.get_context(self)
```
`MakoTemplateBlockBase` where `get_context()` is implemented is superclass of `MakoModuleDescriptor`.

symbolist commented:
@natabene Thanks! This work is part of the work on Blockstore and @ormsbee will be reviewing it. There are no UX changes expected so we should not need a product review.

symbolist commented:
@bradenmacdonald @ormsbee There are a few test failures to resolve and testing to do but I think the review can be started.

ormsbee commented:
FYI @Colin-Fredericks, @pdpinch

Colin-Fredericks commented:
> FYI @Colin-Fredericks, @pdpinch

Thanks for the heads-up @ormsbee ! Right now it says that the sandbox is still being provisioned, so I'll wait until that's up to start testing.

pdpinch commented:
Thanks @ormsbee and @symbolist -- any hints on what you think needs looking at? I'll definitely check import/export, but is there anything else? Do you think this will touch much on customerresponse behavior?

symbolist commented:
@Colin-Fredericks @pdpinch Thanks for checking in!

> I'll definitely check import/export, but is there anything else? Do you think this will touch much on customerresponse behavior?

@pdpinch We are in the middle of manual testing. Will answer once we are done with it and are clear about all the aspects.

symbolist commented:
@pomegranited Since you have done most of the work on the conversion will you time to do a code review as well this sprint? I am particularly interest in the import/export code flow and if there are any subtle issues that are being missed here.

symbolist commented:
@pomegranited Thank you for your review!

The problems with the advanced CAPA problems seems to be present in master too. I am seeing 404s for CAPA file fetches like: http://localhost:18010/static/js/capa/jsmolcalc/jsmolcalc.nocache.js I suspect it is because the studio static files are being nested under the /static/studio/ path now? Though this seems to be an issue only in devstack and not in production.

I also did some testing of import+export of the demo course and a few other courses I have access to (1. import course on master 2. export course on master 3. commit to a git repo. 4. import it on this branch 5. export it on this branch) I am seeing the same output on master and this branch. I think there was also some recent change which is impacting the whitespace in the output causing a lot of noise.

`TypeError: Value stored in a List must be None or a list, found <type 'str'>`

This error exists on master too.

Having said so I had forgotten to add `XmlMixin` that I had created to the `ProblemBlock`. I do not think it changes any behavior for CAPA but just want to make sure all the methods are the same.

Will add notes about the remaining issues later.

Colin-Fredericks commented:
I've got a metric ton of advanced problems and would be happy to test them out. Any idea when the sandbox will be ready?

symbolist commented:
@Colin-Fredericks Fantastic! There is one last bug that we need to resolve. Just waiting for that to happen before we provision the final sandbox for testing. Hopefully, late next week.

symbolist commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -1066,32 +1106,41 @@ class XModuleDescriptor(HTMLSnippet, ResourceTemplates, XModuleMixin):
+                # This init happened the first time an XModule attribute was accessed via the proxy on the
+                # XModuleDescriptor. So we do it here instead.
+                self.init_only_for_xmodules_converted_to_xblocks()
+        return super(XModuleDescriptorToXBlockMixin, self).__getattribute__(name)
```
@ormsbee @cpennington @bradenmacdonald There is one last issue for which I am still trying to come up with a clean solution. For XModuleDescriptors/XModules, the XModule object is initialized (and save() is called) when an XModule property is accessed on the Descriptor in [_xmodule()](https://github.com/edx/edx-platform/blob/release-2019-04-02-10.13/common/lib/xmodule/xmodule/x_module.py#L1199-L1233). 

I am trying to keep the behavior exactly the same. Would this be the right equivalent or is there a better option?

(For reviewers, the first solution of doing it in `bind_for_student()` wasn't the right one. This is attempt 2.)

cpennington replied:
My recollection of how I left this was that I was intending for true XBlocks to not have the `_xmodule` property. https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1885-L1895 suggests the same.

Which behavior are you trying to preserve? The extra call to `.save`?

symbolist replied:
@cpennington Here is an overview of approaches and issues:

1. `CapaModule` had a `[init](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/capa_base.py#L237-L305)` which needs attributes available on the `ModuleSystem`. This was called in `_xmodule()` so it worked and we cannot call this code in `ProblemBlock` `init()`.
2. I tried calling it in `bind_for_student()` after the `xmodule_runtime` is set but that gets called [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/views/index.py#L179) in `CoursewareIndex` when trying to get the children before `self._prefetch_and_bind_section()` which causes a query to Mongo for each `ProblemBlock`. Unless we can do the caching earlier, this init has to happen later.
3. This is option 3.
4. We could delay it even further by overriding `__getattribute__` in `ProblemBlock` instead and doing the init whenever one of the attributes which are touched by `init` is called. But then we would have to do this for each XModule we convert which uses this pattern. I see at least Video and Seq use it.

As for the `save()` call I am not sure if we need it. Do we? In my earlier solutions, there were some issues with `student_view` not saving/loading the correct state that I did not investigate further because I decided to change to approach. But I am trying to keep the logic as same as possible to be on the safe side.

symbolist replied:
Stacktrace for 2 above.

```
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views/index.py", line 153, in get
    return self.render(request)
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views/index.py", line 179, in render
    if self.course.has_children_at_depth(CONTENT_DEPTH):
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in has_children_at_depth
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in <genexpr>
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in has_children_at_depth
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in <genexpr>
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 517, in has_children_at_depth
    return bool(self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 558, in get_children
    in super(XModuleMixin, self).get_children(usage_id_filter)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 407, in get_children
    if usage_id_filter is None or usage_id_filter(usage_id)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 568, in get_child
    child = super(XModuleMixin, self).get_child(usage_id)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 393, in get_child
    child_block = self.runtime.get_block(usage_id, for_parent=self)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1904, in get_block
    return self.get_module(self.descriptor_runtime.get_block(block_id, for_parent=for_parent))
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 531, in inner_get_module
    will_recheck_access=True,
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 894, in get_module_for_descriptor_internal
    partial(LmsFieldData, student_data=student_data),
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 721, in bind_for_student
    self.init_only_for_xmodules_converted_to_xblocks()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 376, in init_only_for_xmodules_converted_to_xblocks
    self._init_only_for_xmodules_converted_to_xblocks()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/capa_base.py", line 131, in _init_only_for_xmodules_converted_to_xblocks
    self.lcp = self.new_lcp(self.get_state_for_lcp())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/capa_base.py", line 201, in new_lcp
    text = self.data
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/fields.py", line 518, in __get__
    if field_data.has(xblock, self.name):
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/field_data.py", line 176, in has
    return self._field_data(block, name).has(block, name)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/field_data.py", line 200, in has
    return self._source.has(block, name)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 219, in has
    return self._kvs.has(self._key(block, name))
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/split_mongo_kvs.py", line 146, in has
    self._load_definition()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/split_mongo_kvs.py", line 182, in _load_definition
    persisted_definition = self._definition.fetch()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/definition_lazy_loader.py", line 31, in fetch
    definition = self.modulestore.get_definition(self.course_key, self.definition_locator.definition_id)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/split.py", line 402, in get_definition
    definition = self.db_connection.get_definition(definition_guid, course_key)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py", line 536, in get_definition
    definition = self.definitions.find_one({'_id': key})
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mongodb_proxy.py", line 53, in wrapper
    return func(*args, **kwargs)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/collection.py", line 1084, in find_one
    for result in cursor.limit(-1):
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/cursor.py", line 1149, in next
    if len(self.__data) or self._refresh():
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/cursor.py", line 1081, in _refresh
    self.__codec_options.uuid_representation))
```

symbolist replied:
I took a look at `_get_student_block()` too. Given that it is only called from `render()` I do not think it would be the right hook because `render()` should only be called when rendering the views and not for other cases like grading, handlers etc.

cpennington replied:
So, one of the intentional design choices around XBlock was that they not have anything in `__init__`, because any given XBlock could be instantiated in a bunch of different contexts (some of which only have access to a subset of the available data). You're running into exactly this problem.

Going through the various things in `CapaMixin.__init__`:
1. `self.close_date` could be a `property` or a `cached_property`, rather than an attribute set during `__init__`
2. `choose_new_seed` could probably be run at when `new_lcp` is called, since I think that's the only place it's used.
3. `lcp` could be converted to a `cached_property` so that it can be generated on first use, rather than during `__init__`, but then not re-generated on future uses.
4. I'm not sure about the comment regarding `openendedresponse`. Is that still in use in the platform?

ormsbee replied:
`openendedresponse` was an input type associated with ORA1, and no longer exists in the platform.

symbolist replied:
@cpennington Thanks! I was hoping to do that in a later round (together with other converted XBlocks) to reduce risk but I'll take a look at getting rid of the `__init__`.

symbolist commented on common/lib/xmodule/xmodule/tests/test_xblock_wrappers.py:
```diff
@@ -49,7 +48,6 @@
 # TODO: Add more types of sample data
 LEAF_XMODULES = {
     AnnotatableDescriptor: [{}],
-    CapaDescriptor: [{}],
```
This are 4 related tests:

1. `TestStudentView`: Not needed since `student_view` is implement by `ProblemBlock`.
2. `TestStudioView`: Not needed since `studio_view` is implement by `ProblemBlock`.
3. `TestPublicView`: `ProblemBlock` uses the default placeholder `public_view` so not that important a test.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -85,224 +105,44 @@ def default(self, obj):  # pylint: disable=method-hidden
-        # Need the problem location in openendedresponse to send out.  Adding
-        # it to the system here seems like the least clunky way to get it
-        # there.
-        self.runtime.set('location', text_type(self.location))
```
Was needed by ORA1. No longer required.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -85,224 +105,44 @@ def default(self, obj):  # pylint: disable=method-hidden
-            # every time the module is loaded.
-            # So we set the seed ONLY when there is not one set already
-            if self.seed is None:
-                self.seed = self.lcp.seed
```
At the time this was added in https://github.com/edx/edx-platform/commit/fb9760244262073384c90ec0de29e737455c5ce8, seed was not generated for all cases in `CapaModule`: https://github.com/edx/edx-platform/blob/fb9760244262073384c90ec0de29e737455c5ce8/common/lib/xmodule/xmodule/capa_module.py#L137-L142

And could be generated in: https://github.com/edx/edx-platform/blob/fb9760244262073384c90ec0de29e737455c5ce8/common/lib/capa/capa/capa_problem.py#L119-L121

But all cases were moved to `CapaModule` in https://github.com/edx/edx-platform/commit/477fe670dd1b30ae97de3258cc8710255941c4a4#diff-54270f54a6e2a6fcbb04e74d13e191a3. 

So we not longer need to do this.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -85,224 +105,44 @@ def default(self, obj):  # pylint: disable=method-hidden
-                # add extra info and raise
-                raise Exception(msg), None, sys.exc_info()[2]
-
-            self.set_state_from_lcp()
```
This line only got executed if `self.runtime.DEBUG` was `True` so no longer needed.

symbolist commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -234,75 +267,40 @@ class CapaMixin(ScorableXBlockMixin, CapaFields):
     """
         Core logic for Capa Problem, which can be used by XModules or XBlocks.
     """
-    def __init__(self, *args, **kwargs):
```
@cpennington @ormsbee Okay, I have removed `init()` and moved the code to properties.

symbolist replied:
@ormsbee Whenever you are able to get to this, the one area I would like a deeper review on is the seed generation logic.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -6,48 +6,122 @@
+        # Since bind_for_student clears out field_data_cache we also reset self.lcp
+        # which will be regenerated on next access.
+        if 'lcp' in self.__dict__:
+            del self.__dict__['lcp']
```
@cpennington I had to do this to fix a few tests:

1. In `openedx.features.content_type_gating.tests.test_access.TestProblemTypeAccess`, `self.score` get reset to `None` when `bind_for_student()` is called again on the block but is not regenerated because `lcp` already exists.
2. There were failures in `lms.djangoapps.instructor_task.tests.test_integration.TestRescoringTask` related to unexpected scores.

Not sure how long Jenkins keeps the test run data but it is at: https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/3940/#showFailuresLink

symbolist replied:
Sorry, I don't have the stack traces right now. Will get to this tomorrow.

symbolist replied:
When `bind_for_student()` is called, a new `xmodule_runtime` instance is set on the descriptor. Next time a module property is accessed, [this](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1207) would be `None` and a new `CapaModule` object gets created.

(I'll improve the docs).

symbolist replied:
@ormsbee I have improved the comment.

`hasattr` executes the property block so that would cause an unnecessary init of the lcp object.

natabene commented:
@symbolist Do you think this is ready for our review?

pdpinch commented:
Done, thanks.

> Can you please create an account

Colin-Fredericks commented:
Done, under colin.fredericks@gmail.com. Thanks!

symbolist commented:
@pdpinch @Colin-Fredericks Thanks! I have given both of you staff access on the sandbox. You may need to sign out and sign in again to see that reflected (FYI @pdpinch there was a typo in your address; it ended with a `@mite.du`. Have fixed it.)

A few notes about testing:
- Overall, there should not be any changes in functionality of CAPA.
- I have access to a few production course archives and have noticed that importing them in when on the `master` branch and exporting out results in differences from the original. Since we want to verify that only the code in this PR does not introduce any changes to CAPA export/import, ideally the testing should be done with course archives which were exported out from an edX app instance which is on a recent commit of edx-platform.
- When creating a course if you use the same course key as that of the archive, it would reduce noise when exporting it out for comparison.

natabene commented:
@ormsbee This is ready for your review.

Colin-Fredericks commented:
Testing notes!
All imports are from today.

**Grape Ape** 
Course key: HarvardX+GA001+2015
This is my standard "has everything" course from edge.edx.org. The sandbox was unable to import it, giving an "Error importing course" message. 

**Placeholder Course Name** (yes, that's its name)
Course key: HarvardX+HX101+1T2019
This was created by my course template builder, and consists mostly of the HarvardX boilerplate course and a blank skeleton. Import and export were both successful. Problems were generated in a "problem" folder (hooray!). Some videos were generated in a "video" folder, some were not. On examination, it turns out that this revealed a bug in my template builder! Export is only producing items that are in the course structure, which is a known bug, so hooray for bug-for-bug compatibility.

**Justice**
Course key: HarvardX+ER22.1x+1T2019
One of our traditional high-enrollment courses that's always in rerun. Includes split_test folder. Imported and exported fine. The number of items in the problem and split_test folders match, but the new video folder has one fewer item (86636740998b45d291c751c81bda9404.xml). The vertical for this video (vert_l24_poll_new.xml) exists in both courses, but is different in the export from edx.org and the export from the sandbox. The vertical in the sandbox does not contain the video. Checking inside the course, it appears that the video was stripped out during import: https://studio.pr20023.sandbox.opencraft.hosting/container/block-v1:HarvardX+ER22.1x+1T2019+type@vertical+block@vert_l24_poll_new  Every other video in the course seems fine, though I didn't go through page-by-page to check.

**Contracts**
Course key: HarvardX+HLS2X+1T2019a
Another perpetual rerun. This has a poll_question folder and some ORA questions. Imported and exported fine. No differences found, except that the old one has an empty "drafts" folder and the new one doesn't, which seems like an acceptable difference to me.

symbolist commented:
@Colin-Fredericks Thank you so much for the testing!

> Grape Ape
> Course key: HarvardX+GA001+2015

The import is failing due to the `ubcpi` xblock not being installed. I have installed it now. Can you give it a try again?

> Justice
> Course key: HarvardX+ER22.1x+1T2019
> The vertical for this video (vert_l24_poll_new.xml) exists in both courses, but is different in the export from edx.org and the export from the sandbox. The vertical in the sandbox does not contain the video. Checking inside the course, it appears that the video was stripped out during import: https://studio.pr20023.sandbox.opencraft.hosting/container/block-v1:HarvardX+ER22.1x+1T2019+type@vertical+block@vert_l24_poll_new Every other video in the course seems fine, though I didn't go through page-by-page to check.

We haven't made any changes to Video XBlock so it shouldn't be impacted by this PR. But if you can add the archive to the Tarballs for Testing Google Drive folder you shared earlier, I can double check.

Colin-Fredericks commented:
> The import is failing due to the `ubcpi` xblock not being installed. I have installed it now. Can you give it a try again?
> 

Sure, though I should mention that there's currently a known bug with exporting and reimporting that particular XBlock.

I got another "Error importing course". My guess is that it's yet another XBlock. Here's the complete Advanced Module List in Grape Ape:

```
[
    "openassessment",
    "word_cloud",
    "lti_consumer",
    "split_test",
    "library_content",
    "recommender",
    "poll",
    "survey",
    "ubcpi",
    "drag-and-drop-v2",
    "done",
    "problem-builder",
    "step-builder"
]
```

We've also got "Add Unsupported Problems and Tools" set to "true". 

> We haven't made any changes to Video XBlock so it shouldn't be impacted by this PR. But if you can add the archive to the Tarballs for Testing Google Drive folder you shared earlier, I can double check.

Done! Thanks.

symbolist commented:
I see `ItemNotFoundError: BlockKey(type=u'videoannotation', id=u'b28e4644561a4eee872c9abd3e814306')` in the logs now. Though the [import page](https://studio.pr20023.sandbox.opencraft.hosting/import/course-v1:HarvardX+GA001+2015) shows the import succeeded?

According to https://github.com/edx/edx-platform/pull/17299/files#diff-ce4c2314083b2185f4c20cd63ceef09cL31, the three annotation modules were removed. Maybe removing references to them from the archive should do the trick?

symbolist commented:
@Colin-Fredericks ^^

Colin-Fredericks commented:
@symbolist  Feel free to remove whatever fails; I'm just a bottleneck in this process since the Import page fails without giving me any useful information.

ormsbee commented:
jenkins run python

symbolist commented:
@Colin-Fredericks Sorry, I missed your earlier message.

> Justice
Course key: HarvardX+ER22.1x+1T2019
The vertical for this video (vert_l24_poll_new.xml) exists in both courses, but is different in the export from edx.org and the export from the sandbox. The vertical in the sandbox does not contain the video. Checking inside the course, it appears that the video was stripped out during import: https://studio.pr20023.sandbox.opencraft.hosting/container/block-v1:HarvardX+ER22.1x+1T2019+type@vertical+block@vert_l24_poll_new Every other video in the course seems fine, though I didn't go through page-by-page to check.

Hmm. I didn't actually check the link earlier and now after re-importing the course, I do see the video there. The only difference I see is the change in name of an srt file for a different video but that has likely to do something with val.

> Course key: HarvardX+GA001+2015

The import succeeded after I removed the annotation modules. On export I see quite a few white-space differences and comments have been removed but otherwise the xml looks the same. Is the archive you put in the shared folder an export from edx.org or from a hand-edited repo?

symbolist commented:
@pdpinch Have you done the testing you were interested in?

Colin-Fredericks commented:
> Hmm. I didn't actually check the link earlier and now after re-importing the course, I do see the video there. The only difference I see is the change in name of an srt file for a different video but that has likely to do something with val.
> 

Ok; as long as it's working now.

> > Course key: HarvardX+GA001+2015
> 
> The import succeeded after I removed the annotation modules. On export I see quite a few white-space differences and comments have been removed but otherwise the xml looks the same. Is the archive you put in the shared folder an export from edx.org or from a hand-edited repo?

This course has been exported, edited, re-imported, re-exported, processed with multiple Python scripts, re-re-imported, and generally been through the wringer. :) There's a reason we use this as our standard test case.

Whitespace differences are fine (as long as they're not inside &lt;pre&gt; tags, &lt;code&gt; tags, or &lt;script&gt; tags), but comments should not be stripped out - those are important.

pdpinch commented:
Not Yet. Will try to do it tonight (I’m in EDT)

> @pdpinch <https://github.com/pdpinch> Have you done the testing you were interested in?

symbolist commented:
@Colin-Fredericks 

> This course has been exported, edited, re-imported, re-exported, processed with multiple Python scripts, re-re-imported, and generally been through the wringer. :) There's a reason we use this as our standard test case.

Content in modulestore is not stored as XML but as Mongo documents so it is fine if it has been edited in different ways outside. My question was specifically if the archive has been edited in any way after its last export from edx.org? If a course is exported from edx.org and imported-exported from this stage server that means the code in this PR does not impact import-export and that is what we want to check.

> Whitespace differences are fine (as long as they're not inside <pre> tags, <code> tags, or <script> tags), but comments should not be stripped out - those are important.

I don't think any XML comments would remain after a course has been imported in Mongo and exported out.

@ormsbee Is that correct?

ormsbee commented:
@symbolist:

I believe that @Colin-Fredericks is referring to XML comments in the Capa problem definition (as opposed to in the broader OLX)? In which case, that is definitely preserved.

Colin-Fredericks commented:
> Content in modulestore is not stored as XML but as Mongo documents so it is fine if it has been edited in different ways outside. My question was specifically if the archive has been edited in any way after its last export from edx.org? If a course is exported from edx.org and imported-exported from this stage server that means the code in this PR does not impact import-export and that is what we want to check.
> 

Ah, ok. No, I exported it from Edge and imported it directly to the sandbox. No touch-up inbetween.

> @symbolist:
> 
> I believe that @Colin-Fredericks is referring to XML comments in the Capa problem definition (as opposed to in the broader OLX)? In which case, that is definitely preserved.

Yeah, comments in problem files definitely remain after a course is exported and re-imported. I don't know of other places that I can actually insert any comments in Studio, so for most folks those will be the only places - though, as always, MIT edits XML directly and might have some in their course structures.

symbolist commented:
@ormsbee @Colin-Fredericks Okay, let me take a second look at HarvardX+GA001+2015.

pdpinch commented:
@symbolist I tried an import round-trip of our devops course and [@jolyonb's mitx-grading-library demo course](https://github.com/mitodl/mitx-grading-library/tree/master/course). Problems in both courses seemsed to function properly after import. On export, they had a few changes to whitespace and HTML quirks, but nothing unexpected. 

I give it a :+1: -- amazing work.

symbolist commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -6,48 +6,122 @@
+        """
+        Return the student view.
+        """
+        self.lcp
```
@ormsbee Added comment to explain this as well.

symbolist commented:
@pdpinch Thanks for your help in testing this out!

@ormsbee @Colin-Fredericks 

> I believe that @Colin-Fredericks is referring to XML comments in the Capa problem definition (as opposed to in the broader OLX)? In which case, that is definitely preserved.

It does not look like this is the case anymore. To double check I [added some comments](https://gist.github.com/symbolist/383132886f8af7da5708e8fb7baa035e) to a capa problem in the demo course and imported it to a sandbox (which is on master) and then re-exported it. The export no longer has the comments.

I also tested `HarvardX+ER22.1x+1T2019` on both master and this branch and the import-export roundtrip has no differences.

@ormsbee I have improved the comments, squashed and rebased.

ormsbee commented:
@symbolist: I am concerned if we're now stripping comments from capa problems (if this behavior wasn't already present). It's not a commonly used feature, but it's a feature used by our power-users who are deeply invested in their problem authoring.

Also, please supplement your commit message with all the awesome context in your PR message.

ormsbee commented:
Confirmed that current behavior mirrors what's on prod today. So as soon as the commit message gets fleshed out with details, I'll press the merge button.

symbolist commented:
jenkins run python

edx-webhook commented:
@symbolist 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/20023&entry.1484655318=https://github.com/symbolist&entry.752974735=2019-03-19+14:41&entry.1917517419=2019-05-01+16:18&entry.2133058324=Yes) so we can improve your experience in the future.

bradenmacdonald commented:
\o/

azarembok commented:
@symbolist (cc: @ormsbee) we are investigating an issue with problem editing that may be related to this change.  If a user creates an advanced problem that contains python code, and the user-supplied python code gets an exception, that exception causes the problem to be uneditable.  I am attaching a screenshot and a stack trace that shows where the error occurs.  Any suggestions on a fix would be welcome.  Thanks in advance!

![missing edit component](https://user-images.githubusercontent.com/12628682/60914548-c0ae1780-a257-11e9-95b5-99f4c67990a6.png)

[stack.txt](https://github.com/edx/edx-platform/files/3374429/stack.txt)

Colin-Fredericks commented:
In brief defense of this PR, that error has _always_ been associated with ironically uneditable pages. It just seems easier to get that error now.

symbolist commented:
Hi @azarembok, when CAPA was an XModule, the `lcp` init should have happened in this [method](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1271). As you can see there is a broad exception catcher which replaces the instance with an `error_descriptor_class` which gets rendered instead.

However, as far as I understand we don't have a similar hook for XBlocks to be able to initialize `lcp` in so we are doing it lazily latter (in this case it is happening when `student_view()` is rendered). Do we know what happens when viewing this problem in LMS? I am assuming there is a broad exception catcher somewhere in case `student_view()` of any xblock raises an exception.

If we can't fix it there then the other option I can think of is to catch [this](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/capa_base.py#L299) `reraise` in the CAPA `student_view()` and return an error message instead.

azarembok commented:
Thanks very much for your input @symbolist .  To answer your question above about behavior in the LMS, it was failing with a similar exception.  I have fixed the issue with this PR: https://github.com/edx/edx-platform/pull/21093
